File: src/services/echoApi.ts
import { invoke } from '@tauri-apps/api/tauri';

export interface AddressWithRest
{
  addr: string,
  rest: string | null,
}

export interface Message
{
  address: AddressWithRest,
  timestamp: number,
  content: string,
}

export interface MessageHistory
{
  messages: Array<Message>,
}
export interface LiteCoinAddress
{
  address: AddressWithRest,
  amount: number,
}

export interface LiteCoinAddresses
{
  addresses: Array<LiteCoinAddress>,
}

class EchoApi
{
  private userAddress?: string;

  constructor(userAddress?: string)
  {
    const userAddress_ = userAddress || sessionStorage.getItem('userAddress');
    if (userAddress_)
      this.setUserAddress(userAddress_);
  }

  public setUserAddress(userAddress: string)
  {
    this.userAddress = userAddress;
    sessionStorage.setItem('userAddress', userAddress);
  }

  public sendMessage(message: string): Promise<boolean>
  {
    if (!this.userAddress)
      throw new Error('InvalidUserAddress');
    return invoke('send_message', { user_address: this.userAddress, message });
  }

  public async getMessages(): Promise<MessageHistory>
  {
    return invoke('get_messages');
  }

  public async getUserAddresses(): Promise<LiteCoinAddresses>
  {
    return invoke('get_users_addresses');
  }
}

export default new EchoApi();

File: src-tauri/src/echo_api.rs
use tauri;

#[derive(serde::Serialize)]
pub struct AddressWithRest {
  rest: Option<String>,
  addr: String,
}

#[derive(serde::Serialize)]
pub struct LiteCoinAddress {
  amount: f32,
  address: AddressWithRest,
}

#[derive(serde::Serialize)]
pub struct LiteCoinAddresses {
  addresses: Vec<LiteCoinAddress>,
}

#[tauri::command]
pub async fn get_users_addresses() -> Result<LiteCoinAddresses, String> {
  let mut addresses: Vec<LiteCoinAddress> = vec![];
  for i in 1..10 {
    if i%2 == 0 {
      addresses.push(LiteCoinAddress {
        amount: (i as f32) / 10.0,
        address: AddressWithRest {
          addr: format!("Addr {}", i),
          rest: Some("Jean Michel".to_string()),
        }
      })
    } else {
      addresses.push(LiteCoinAddress {
        amount: (i as f32) / 10.0,
        address: AddressWithRest {
          addr: format!("Addr {}", i),
          rest: None,
        }
      })
    }
  }
  Ok(LiteCoinAddresses {
    addresses
  })
}

#[tauri::command]
pub async fn send_message(user_address: String, message: String) -> Result<bool, String> {
  println!("{user_address}");
  println!("{message}");
  Ok(true)
}

#[derive(serde::Serialize)]
struct Message {
  address: AddressWithRest,
  timestamp: u128,
  content: String,
}

#[derive(serde::Serialize)]
pub struct MessageHistory {
  messages: Vec<Message>,
}

#[tauri::command]
pub async fn get_messages() -> Result<MessageHistory, String> {
  let mut messages: Vec<Message> = vec![];
  for i in 1..10 {
    messages.push(Message {
      address: AddressWithRest {
        addr: format!("Address {}", i),
        rest: Some("Patoche".to_string()),
      },
      timestamp: i,
      content: format!("Message {}", i),
    });
  }
  Ok(MessageHistory {
    messages
  })
}